File: Entity/InvoiceItem.php
<?php
/**
 * Author: Till Wegmüller
 * Date: 9/24/14
 * Dime
 */

namespace Dime\InvoiceBundle\Entity;


use Dime\TimetrackerBundle\Entity\Activity;
use Dime\TimetrackerBundle\Entity\Entity;
use Dime\TimetrackerBundle\Model\DimeEntityInterface;
use Doctrine\ORM\Mapping as ORM;
use JMS\Serializer\Annotation as JMS;
use Knp\JsonSchemaBundle\Annotations as Json;
use Money\Money;

/**
 * Class InvoiceItem
 * @package Dime\InvoiceBundle\Entity
 * @ORM\Entity(repositoryClass="Dime\InvoiceBundle\Entity\InvoiceItemRepository")
 * @ORM\Table(name="invoice_items")
 * @Json\Schema("invoiceItems")
 */
class InvoiceItem extends Entity implements DimeEntityInterface
{
	/**
	 * @var Invoice
	 * @ORM\ManyToOne(targetEntity="Dime\InvoiceBundle\Entity\Invoice", inversedBy="items")
	 * @JMS\MaxDepth(1)
	 */
	protected $invoice;

	/**
	 * @var string
	 * @ORM\Column(type="string")
	 */
	protected $name;

	/**
	 * @var int
	 * @ORM\Column(name="rate_value", type="money", nullable=true)
	 * @JMS\SerializedName("rateValue")
	 * @JMS\Type(name="Money")
	 */
	protected $rateValue;

	/**
	 * @var string
	 * @ORM\Column(type="string")
	 * @JMS\SerializedName("rateUnit")
	 */
	protected $rateUnit;

	/**
	 * @ORM\Column(type="decimal", scale=3, precision=10, nullable=true)
	 */
	protected $vat;

	/**
	 * @var string
	 * @ORM\Column(type="string")
	 */
	protected $amount;

	/**
	 * @var Money
	 * @JMS\Type(name="Money")
	 * @ORM\Column(name="total", type="money")
	 */
	protected $total;

	/**
	 * @var Activity
	 * @ORM\ManyToOne(targetEntity="Dime\TimetrackerBundle\Entity\Activity", )
	 */
	protected $activity;

	/**
	 * @param Activity $activity
	 *
	 * @return $this
	 */
	public function setFromActivity(Activity $activity)
	{
		$this->name         = $activity->getName();
		$this->rateValue    = $activity->getRateValue();
		$this->amount       = $activity->serializeValue();
		$this->rateUnit     = $activity->getRateUnit();
		$this->vat          = $activity->getVat();
		$this->total        = $activity->getCharge();
		$this->activity     = $activity;
		return $this;
	}

	/**
	 * @JMS\VirtualProperty
	 * @JMS\SerializedName("calculatedVAT")
	 * @JMS\Type(name="Money")
	 * @return Money
	 */
	public function getCalculatedVAT()
	{
		if($this->rateValue instanceof Money && is_numeric($this->amount) && is_numeric($this->vat))
			return $this->rateValue->multiply($this->amount)->multiply($this->vat);
		else
			return null;
	}

	/**
	 * @return Money
	 */
	public function getTotal()
	{
		return $this->total;
	}

	/**
	 * @param Money $total
	 *
	 * @return $this
	 */
	public function setTotal($total)
	{
		$this->total = $total;
		return $this;
	}

	/**
	 * @return Invoice
	 */
	public function getInvoice()
	{
		return $this->invoice;
	}

	/**
	 * @param Invoice $invoice
	 *
	 * @return $this
	 */
	public function setInvoice($invoice)
	{
		$this->invoice = $invoice;
		return $this;
	}

	/**
	 * @return string
	 */
	public function getRateUnit()
	{
		return $this->rateUnit;
	}

	/**
	 * @param string $rateUnit
	 *
	 * @return $this
	 */
	public function setRateUnit($rateUnit)
	{
		$this->rateUnit = $rateUnit;
		return $this;
	}

	/**
	 * @return int
	 */
	public function getRateValue()
	{
		return $this->rateValue;
	}

	/**
	 * @param int $rateValue
	 *
	 * @return $this
	 */
	public function setRateValue($rateValue)
	{
		$this->rateValue = $rateValue;
		return $this;
	}

	/**
	 * @return string
	 */
	public function getName()
	{
		return $this->name;
	}

	/**
	 * @param string $name
	 *
	 * @return $this
	 */
	public function setName($name)
	{
		$this->name = $name;
		return $this;
	}

	/**
	 * @return float
	 */
	public function getAmount()
	{
		return $this->amount;
	}

	/**
	 * @param float $amount
	 *
	 * @return $this
	 */
	public function setAmount($amount)
	{
		$this->amount = $amount;
		return $this;
	}

	/**
	 * @return mixed
	 */
	public function getVat()
	{
		return $this->vat;
	}

	/**
	 * @param mixed $vat
	 *
	 * @return $this
	 */
	public function setVat($vat)
	{
		$this->vat = $vat;
		return $this;
	}

	/**
	 * @return Activity
	 */
	public function getActivity()
	{
		return $this->activity;
	}

	/**
	 * @param Activity $activity
	 *
	 * @return $this
	 */
	public function setActivity($activity)
	{
		$this->activity = $activity;
		return $this;
	}


} 

File: Resources/public/lib/model/entities/InvoiceItem.dart
part of dime_entity;

class InvoiceItem extends Entity {
  init({Map<String, dynamic> params: const {}}) {
    if (!params.containsKey('name')) {
      params['name'] = 'New Item';
    }
    if (params.containsKey('invoice')) {
      params['invoice'] = new Invoice()
        ..id = params['invoice'];
    }
    super.init(params: params);
  }

  InvoiceItem();

  InvoiceItem.clone(InvoiceItem original): super.clone(original){
    this.name = original.name;
    this.amount = original.amount;
    this.rateValue = original.rateValue;
    this.rateUnit = original.rateUnit;
    this.activity = original.activity;
    this.vat = original.vat;
    this.invoice = original.invoice;
  }

  InvoiceItem.fromMap(Map<String, dynamic> map): super.fromMap(map);

  newObj() {
    return new InvoiceItem();
  }

  dynamic Get(String property) {
    var val = super.Get(property);
    if (val == null) {
      switch (property) {
        case 'rateValue':
          return this.rateValue;
        case 'rateUnit':
          return this.rateUnit;
        case 'amount':
          return this.amount;
        case 'activity':
          return this.activity;
        case 'vat':
          return this.vat;
        case 'invoice':
          return this.invoice;
        case 'calculatedVAT':
          return this.calculatedVAT;
        case 'total':
          return this.total;
        default:
          break;
      }
    }
    return val;
  }

  void Set(String property, var value) {
    switch (property) {
      case 'rateValue':
        this.rateValue = value;
        break;
      case 'rateUnit':
        this.rateUnit = value;
        break;
      case 'amount':
        this.amount = value;
        break;
      case 'activity':
        this.activity = value is Entity ? value : new Activity.fromMap(value);
        break;
      case 'vat':
        this.vat = value;
        break;
      case 'invoice':
        this.invoice = value is Entity ? value : new Invoice.fromMap(value);
        break;
      case 'calculatedVAT':
        this.calculatedVAT = value;
        break;
      case 'total':
        this.total = value;
        break;
      default:
        super.Set(property, value);
        break;
    }
  }

  static List<InvoiceItem> listFromMap(List content) {
    List<InvoiceItem> array = new List<InvoiceItem>();
    for (var element in content) {
      array.add(new InvoiceItem.fromMap(element));
    }
    return array;
  }

  String type = 'invoiceitems';
  Invoice invoice;
  String rateValue;
  String rateUnit;
  String calculatedVAT;
  dynamic amount;
  String total;
  Activity activity;
  double vat;
}